src/main/java/com/iota/iri/service/garbagecollector/MilestonePrunerJob.java

```diff
@@ -0,0 +1,302 @@
+        }
+    }
+
+    /**
+     * This method takes care of cleaning up a single milestone and all of its transactions and performs the actual
+     * database operations.
+     *
+     * This method performs the deletions in an atomic way, which means that either the full processing succeeds or
+     * fails. It does that by iterating through all the transactions that belong to the current milestone and first
+     * collecting them in a List of items to delete. Once all transactions where found we issue a batchDelete.
+     *
+     * While processing the transactions that are directly or indirectly referenced by the chosen milestone, we also
+     * issue additional {@link UnconfirmedSubtanglePrunerJob}s that remove the orphaned parts of the tangle that branch off
+     * the deleted transactions because they would otherwise loose their connection to the rest of the tangle unless
+     * they are branching off a solid entry point (in which case we wait with the deletion until the solid entry point
+     * becomes irrelevant).
+     *
+     * After removing the entries from the database it also removes the entries from the relevant runtime caches.
+     *
+     * @throws GarbageCollectorException if something goes wrong while cleaning up the milestone
+     */
+    private void cleanupMilestoneTransactions() throws GarbageCollectorException {
+        try {
+            // collect elements to delete
+            List<Pair<Indexable, ? extends Class<? extends Persistable>>> elementsToDelete = new ArrayList<>();
+            MilestoneViewModel milestoneViewModel = MilestoneViewModel.get(garbageCollector.tangle, currentIndex);
+            if (milestoneViewModel != null) {
+                elementsToDelete.add(new Pair<>(milestoneViewModel.getHash(), Transaction.class));
+                elementsToDelete.add(new Pair<>(new IntegerIndex(milestoneViewModel.index()), Milestone.class));
+                if (!garbageCollector.snapshotManager.getInitialSnapshot().isSolidEntryPoint(milestoneViewModel.getHash())) {
+                    garbageCollector.addJob(new UnconfirmedSubtanglePrunerJob(milestoneViewModel.getHash()));
+                }
+                DAGHelper.get(garbageCollector.tangle).traverseApprovees(
+                milestoneViewModel.getHash(),
+                approvedTransaction -> approvedTransaction.snapshotIndex() >= milestoneViewModel.index(),
+                approvedTransaction -> {
+                    elementsToDelete.add(new Pair<>(approvedTransaction.getHash(), Transaction.class));
+
+                    if (!garbageCollector.snapshotManager.getInitialSnapshot().isSolidEntryPoint(approvedTransaction.getHash())) {
+                        try {
+                            garbageCollector.addJob(new UnconfirmedSubtanglePrunerJob(approvedTransaction.getHash()));
+                        } catch(GarbageCollectorException e) {
+                            throw new RuntimeException(e);
```

Comment:
this is on purpose - since we use the builtin "Consumer" for the DAGHelper we need to wrap our exception into a RuntimeException to be able to throw it.